src/qscan.cpp

```diff
@@ -0,0 +1,874 @@
+    	    delete dev;
+			return 3;
+		default:
+			printf( MSGPREF "%s: unknown error\n",device);
+    	    delete dev;
+			return 3;
+    }
+	printf( MSGPREF "using device '%s': '%s' '%s' '%s'\n", device, dev->ven, dev->dev, dev->fw);
+    //dev->silent = 0;
+//	get_features_list(dev);
+	if (!(flags & FL_DEBUG)) dev->silent++;
+    detect_capabilities(dev);
+	get_buffer_capacity(dev);
+    determine_disc_type(dev);
+    read_capacities(dev);
+
+	detect_vendor_features(dev);
+	get_rw_speeds(dev);
+
+//	if (!(flags & FL_DEBUG)) dev->silent--;
+
+//	if (flags & (FL_INFO | FL_MINFO | FL_SPEED | FL_SPEEDS))
+	detect_speeds(dev);
+//	test_dma_speed(dev);
+
+    if (flags & FL_INFO) {
+		printf(IDEV "Device: '%s'\n", device);
+		printf(IDEV "Vendor: '%s'\n",dev->ven);
+		printf(IDEV "Model : '%s'\n",dev->dev);
+		printf(IDEV "F/W   : '%s'\n",dev->fw);
+		if (isPlextor(dev)) {
+			plextor_get_TLA(dev);
+			printf(IDEV "TLA#  : %s\n",dev->TLA);
+		} else {
+			printf(IDEV "TLA#  : N/A \n");
+		}
+		printf(IDEV "Buffer: %d kB\n", dev->buffer_size);
+		printf(IDEV "S/N   : %s\n", dev->serial);
+		printf(IDEV "IFace : %s\n", dev->iface);
+		printf(IDEV "Loader: %s\n", loader_list[dev->loader_id]);
+
+		if (dev->life.ok) {
+			printf(IDEV "Discs loaded: %6d\n", dev->life.dn);
+			printf(IDEV "Drive operating time:\n");
+			printf(IDEV "  CD Rd  : %4d:%02d:%02d\n", dev->life.cr.h, dev->life.cr.m, dev->life.cr.s);
+			printf(IDEV "  CD Wr  : %4d:%02d:%02d\n", dev->life.cw.h, dev->life.cw.m, dev->life.cw.s);
+			if (dev->rd_capabilities & DEVICE_DVD)
+				printf(IDEV "  DVD Rd : %4d:%02d:%02d\n", dev->life.dr.h, dev->life.dr.m, dev->life.dr.s);
+			if (dev->wr_capabilities & DEVICE_DVD)
+				printf(IDEV "  DVD Wr : %4d:%02d:%02d\n", dev->life.dw.h, dev->life.dw.m, dev->life.dw.s);
+		}
+	}
+
+	if (flags & (FL_INFO | FL_MINFO)) {
+		get_rw_speeds(dev);
+
+		printf(IDEV "Current read speed  : %4.1fX  ( %5d kB/s)\n",
+			(float)dev->parms.read_speed_kb / dev->parms.speed_mult,
+			dev->parms.read_speed_kb);
+		if (dev->wr_capabilities) {
+			printf(IDEV "Current write speed : %4.1fX  ( %5d kB/s) \n",
+				(float)dev->parms.write_speed_kb / dev->parms.speed_mult,
+				dev->parms.write_speed_kb);
+		}
+	}
+
+	if (dev->wr_capabilities) {
+		check_write_modes(dev);
+	}
+	if (flags & FL_INFO) {
+		printf(IDEV "Device Generic capabilities : 0x%016LX\n",dev->capabilities);
+		printf(IDEV "Device Read capabilities    : 0x%016LX\n",dev->rd_capabilities);
+		printf(IDEV "Device Write capabilities   : 0x%016LX\n",dev->wr_capabilities);
+	}
+
+	if (dev->wr_capabilities) {
+		printf(IDEV "Supported write modes:");
+		for (int i=0; wr_modes[i].id; i++) {
+			if ((dev->wr_modes & wr_modes[i].id) == wr_modes[i].id) printf(" %s", wr_modes[i].name);
+		}
+		printf("\n");
+	}
+
+
+    if (flags & FL_DINFO) {
+		printf("\nDevice capabilities:\n");	
+		for (int idx=0; capabilities[idx].id || strlen(capabilities[idx].name); idx++) {
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)